src/Features/Core/Portable/MoveDeclarationNearReference/MoveDeclarationNearReferenceCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,96 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Composition;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.MoveDeclarationNearReference
+{
+    [ExportCodeRefactoringProvider(LanguageNames.CSharp, LanguageNames.VisualBasic, Name = PredefinedCodeRefactoringProviderNames.MoveDeclarationNearReference), Shared]
+    [ExtensionOrder(After = PredefinedCodeRefactoringProviderNames.InlineTemporary)]
+    internal sealed class MoveDeclarationNearReferenceCodeRefactoringProvider : CodeRefactoringProvider
+    {
+        public override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+        {
+            var document = context.Document;
+            var textSpan = context.Span;
+            var cancellationToken = context.CancellationToken;
+
+            if (!textSpan.IsEmpty)
+            {
+                return;
+            }
+
+            var statement = await GetLocalDeclarationStatementAsync(document, textSpan, cancellationToken).ConfigureAwait(false);
+            if (statement == null)
+            {
+                return;
+            }
+
+            var syntaxFacts = document.GetLanguageService<ISyntaxFactsService>();
+            var variables = syntaxFacts.GetVariablesOfLocalDeclarationStatement(statement);
+            if (variables.Count != 1)
+            {
+                return;
+            }
+
+            // Don't offer the refactoring inside the initializer for the variable.
+            var initializer = syntaxFacts.GetInitializerOfVariableDeclarator(variables[0]);
+            var applicableSpan = initializer == null
+                ? statement.Span
+                : TextSpan.FromBounds(statement.SpanStart, initializer.SpanStart);
+
+            if (!applicableSpan.IntersectsWith(textSpan.Start))
+            {
+                return;
+            }
```

Comment:
this is the refactoring side of things.  So it is the only side concerned with things like positions.  The service side concerns itself with an actual DeclarationStatement as that is generally what any calling feature woudl want to operate on.

If this is not the split you'd like @mavasani , let me know.

Reply:
That seems like the correct split. Thanks.